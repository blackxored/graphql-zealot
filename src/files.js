// @flow
import {
  always,
  // $FlowFixMe$
  compose,
  cond,
  last,
  memoize,
  merge,
  prop,
  startsWith,
  T,
} from 'ramda';
import mime from 'mime/lite';

const getContentTypeFromExtension = memoize((extension: string) => {
  return mime.getType(extension);
});

/**
 * Return a content type from a filename, uses `mime`.
 *
 * @param filename
 * @returns {string} ContentType
 */
export const getContentTypeFromFilename = (filename: string) => {
  // $FlowFixMe$
  return getContentTypeFromExtension(last(filename.split('.')));
};

/**
 * Convert a content-type to a valid GraphQL scbema type
 *
 * @param file
 * @returns {string} Image, Video, Audio if content type matches, GenericFile otherwise
 */
export const getTypenameForFile = (file: { contentType: string }) => {
  return compose(
    // $FlowFixMe$
    cond([
      [startsWith('image'), always('Image')],
      [startsWith('video'), always('Video')],
      [startsWith('audio'), always('Audio')],
      [T, always('GenericFile')],
    ]),
    prop('contentType'),
  )(file);
};

export const optimisticFileResponse = (file: {
  contentType: string,
  name: string,
  url: string,
}) => {
  if (!file.name) {
    return file;
  }

  const isLocal = !file.url.startsWith('http');

  if (!isLocal) {
    return file;
  }

  const contentType = getContentTypeFromFilename(file.name);

  return merge(
    {
      contentType,
      __typename: getTypenameForFile({ contentType }),
    },
    file,
  );
};
