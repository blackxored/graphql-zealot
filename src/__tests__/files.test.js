import { getTypenameForFile, optimisticFileResponse } from '../files';

describe('files', () => {
  describe('getTypenameForFile', () => {
    it('returns the typename for a file given a content-type', () => {
      expect(getTypenameForFile({ contentType: 'image/jpeg' })).toEqual(
        'Image',
      );
      expect(getTypenameForFile({ contentType: 'audio/mp3' })).toEqual('Audio');
      expect(getTypenameForFile({ contentType: 'video/mp4' })).toEqual('Video');
    });

    it('returns GenericFile for content-types it does not recognize', () => {
      expect(
        getTypenameForFile({ contentType: 'application/octet-stream' }),
      ).toEqual('GenericFile');
      expect(getTypenameForFile({ contentType: 'application/zip' })).toEqual(
        'GenericFile',
      );
      expect(getTypenameForFile({ contentType: 'text/plain' })).toEqual(
        'GenericFile',
      );
    });
  });

  describe('optimisticFileResponse', () => {
    it("returns original file if it doesn't have a name", () => {
      const file = { extra: 'file://xxx' };

      expect(optimisticFileResponse(file)).toBe(file);
    });

    it('returns original file if URL is not remote', () => {
      const file = { name: 'foo.jpg', url: 'http://example.com/foo.jpg' };
      expect(optimisticFileResponse(file)).toBe(file);
    });

    it('returns an optimistic response shape for a file', () => {
      expect(
        optimisticFileResponse({
          name: 'foo.jpg',
          url: 'file://foo.jpg',
          contentType: 'image/jpeg',
          size: 2000,
        }),
      ).toMatchSnapshot();
    });
  });
});
