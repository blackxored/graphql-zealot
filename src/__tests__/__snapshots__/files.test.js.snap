// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`files optimisticFileResponse returns an optimistic response shape for a file 1`] = `
Object {
  "__typename": "Image",
  "contentType": "image/jpeg",
  "name": "foo.jpg",
  "size": 2000,
  "url": "file://foo.jpg",
}
`;
